server/src/main/java/org/elasticsearch/index/shard/StoreRecovery.java

```diff
@@ -408,84 +408,91 @@ private ActionListener<Boolean> recoveryListener(IndexShard indexShard, ActionLi
      * Recovers the state of the shard from the store.
      */
     private void internalRecoverFromStore(IndexShard indexShard, ActionListener<Void> outerListener) {
-        indexShard.preRecovery(outerListener.delegateFailure((listener, ignored) -> ActionRunnable.run(listener, () -> {
-            final RecoveryState recoveryState = indexShard.recoveryState();
-            final boolean indexShouldExists = recoveryState.getRecoverySource().getType() != RecoverySource.Type.EMPTY_STORE;
-            indexShard.prepareForIndexRecovery();
-            SegmentInfos si = null;
-            final Store store = indexShard.store();
-            store.incRef();
-            try {
+        indexShard.preRecovery(outerListener.delegateFailure((listener, ignored) -> new ActionRunnable<>(listener) {
```

Comment:
Creating and then immediately executing a new `ActionRunnable` is really just a try/catch block with extra steps. In this case we could use `delegateFailureAndWrap` instead of `delegateFailure` to get the same effect.